{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Unit testing\n",
    "\n",
    "Test smallest parts of an application in isolation (e.g. units)<br>\n",
    "Good candidates for unit testing: individual classes, modules, or functions<br>\n",
    "Bad candidates for unit testing: an entire application, dependencies across several classes or modules<br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### unittest\n",
    "\n",
    "Python comes with a built-in module called unittest<br>\n",
    "You can write unit tests encapsulated as classes that inherit from unittest.TestCase<br>\n",
    "This inheritance gives you access to many assertion helpers that let you test the behavior of your functions<br>\n",
    "You can run tests by calling unittest.main()<br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Commenting Tests\n",
    "class SomeTests(unittest.TestCase):\n",
    "    def first_test(self):\n",
    "        \"\"\"testing a thing\"\"\"\n",
    "        self.assertEqual(thing(), \"something\")\n",
    "\n",
    "    def second_test(self):\n",
    "        \"\"\"testing another thing\"\"\"\n",
    "        self.assertEqual(another_thing(), \"something else\")\n",
    "\n",
    "        \n",
    "# To see comments, run\n",
    "\n",
    "# python NAME_OF_TEST_FILE.py -v"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "types of esserstions\n",
    "\n",
    "self.assertEqual(x, y)\n",
    "self.assertNotEqual(x, y)\n",
    "self.assertTrue(x)\n",
    "self.assertFalse(x)\n",
    "self.assertIsNone(x)\n",
    "self.assertIsNotNone(x)\n",
    "self.assertIn(x, y)\n",
    "self.assertNotIn(x, y)\n",
    "...and more!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Testing For Error use assertRaises\n",
    "\n",
    "class SomeTests(unittest.TestCase):\n",
    "    def testing_for_error(self):\n",
    "        \"\"\"testing for an error\"\"\"\n",
    "        with self.assertRaises(IndexError):\n",
    "            l = [1,2,3]\n",
    "            l[100]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# unitesting_file.py\n",
    "\n",
    "def eat(food,ishealthy):\n",
    "    if ishealthy:\n",
    "        return \"veg food is healthy\"\n",
    "    return \"lol\"\n",
    "\n",
    "def check_num(num):\n",
    "    if num<2:\n",
    "        return \"this is 1\"\n",
    "    return \"this is 3\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "E\n",
      "======================================================================\n",
      "ERROR: /home/rahish/ (unittest.loader._FailedTest)\n",
      "----------------------------------------------------------------------\n",
      "AttributeError: module '__main__' has no attribute '/home/rahish/'\n",
      "\n",
      "----------------------------------------------------------------------\n",
      "Ran 1 test in 0.025s\n",
      "\n",
      "FAILED (errors=1)\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "True",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[0;31mSystemExit\u001b[0m\u001b[0;31m:\u001b[0m True\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/rahish/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3333: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.\n",
      "  warn(\"To exit: use 'exit', 'quit', or Ctrl-D.\", stacklevel=1)\n"
     ]
    }
   ],
   "source": [
    "# this is a test file\n",
    "# unittest_test_class.py\n",
    "\n",
    "import unittest\n",
    "from unitesting_file import eat,check_num\n",
    "\n",
    "\n",
    "class Testactivity(unittest.TestCase):\n",
    "    \n",
    "    def test_eats(self):\n",
    "        \"\"\"this is veg\"\"\"\n",
    "        self.assertEqual(\n",
    "            eat(\"veg\",ishealthy=True),\n",
    "            \"veg food is healthy\"\n",
    "        )\n",
    "    \n",
    "    def test_eat(self):\n",
    "        \"\"\"this is non veg\"\"\"\n",
    "        self.assertEqual(\n",
    "            eat(\"non-veg\",ishealthy=False),\n",
    "            \"lol\"\n",
    "        )\n",
    "      \n",
    "        \n",
    "    def test_check_num(self):\n",
    "        \"\"\"this is 3\"\"\"\n",
    "        self.assertEqual(\n",
    "        check_num(3),\n",
    "        \"this is 3\"\n",
    "        )\n",
    "        \n",
    "    def test_num(self):\n",
    "        \"\"\"this is 1\"\"\"\n",
    "        self.assertEqual(\n",
    "           check_num(1),\n",
    "            \"this is 1\"\n",
    "        )\n",
    "       \n",
    "if __name__==\"__main__\":\n",
    "    unittest.main()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## after run in terminal result "
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "\n",
    "\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 0 tests in 0.000s\n",
    "\n",
    "OK\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unitesting_file.py\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py\n",
    "E\n",
    "======================================================================\n",
    "ERROR: test_check_sum (__main__.Tests)\n",
    "----------------------------------------------------------------------\n",
    "Traceback (most recent call last):\n",
    "  File \"unittest_testing_class.py\", line 20, in test_check_sum\n",
    "    check_sum(3),\n",
    "NameError: name 'check_sum' is not defined\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 1 test in 0.001s\n",
    "\n",
    "FAILED (errors=1)\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py\n",
    "E\n",
    "======================================================================\n",
    "ERROR: test_check_num (__main__.Tests)\n",
    "----------------------------------------------------------------------\n",
    "Traceback (most recent call last):\n",
    "  File \"unittest_testing_class.py\", line 20, in test_check_num\n",
    "    check_sum(3),\n",
    "NameError: name 'check_sum' is not defined\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 1 test in 0.000s\n",
    "\n",
    "FAILED (errors=1)\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py\n",
    "F\n",
    "======================================================================\n",
    "FAIL: test_check_num (__main__.Tests)\n",
    "----------------------------------------------------------------------\n",
    "Traceback (most recent call last):\n",
    "  File \"unittest_testing_class.py\", line 21, in test_check_num\n",
    "    \"this is 3 and it is wow\"\n",
    "AssertionError: None != 'this is 3 and it is wow'\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 1 test in 0.000s\n",
    "\n",
    "FAILED (failures=1)\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py\n",
    "F\n",
    "======================================================================\n",
    "FAIL: test_check_num (__main__.Tests)\n",
    "----------------------------------------------------------------------\n",
    "Traceback (most recent call last):\n",
    "  File \"unittest_testing_class.py\", line 22, in test_check_num\n",
    "    \"this is 3 and it is wow\"\n",
    "AssertionError: None != 'this is 3 and it is wow'\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 1 test in 0.000s\n",
    "\n",
    "FAILED (failures=1)\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py\n",
    "F\n",
    "======================================================================\n",
    "FAIL: test_check_num (__main__.Tests)\n",
    "----------------------------------------------------------------------\n",
    "Traceback (most recent call last):\n",
    "  File \"unittest_testing_class.py\", line 22, in test_check_num\n",
    "    \"this is 3 and it is wow\"\n",
    "AssertionError: None != 'this is 3 and it is wow'\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 1 test in 0.000s\n",
    "\n",
    "FAILED (failures=1)\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py\n",
    "Traceback (most recent call last):\n",
    "  File \"unittest_testing_class.py\", line 2, in <module>\n",
    "    from unitest_file import eat,check_num\n",
    "ModuleNotFoundError: No module named 'unitest_file'\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py\n",
    "F\n",
    "======================================================================\n",
    "FAIL: test_eat (__main__.Testactivity)\n",
    "----------------------------------------------------------------------\n",
    "Traceback (most recent call last):\n",
    "  File \"unittest_testing_class.py\", line 10, in test_eat\n",
    "    \"veg foog is healthy\"\n",
    "AssertionError: None != 'veg foog is healthy'\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 1 test in 0.000s\n",
    "\n",
    "FAILED (failures=1)\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py\n",
    "F\n",
    "======================================================================\n",
    "FAIL: test_eat (__main__.Testactivity)\n",
    "----------------------------------------------------------------------\n",
    "Traceback (most recent call last):\n",
    "  File \"unittest_testing_class.py\", line 10, in test_eat\n",
    "    \"veg food is healthy\"\n",
    "AssertionError: None != 'veg food is healthy'\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 1 test in 0.000s\n",
    "\n",
    "FAILED (failures=1)\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py\n",
    "F\n",
    "======================================================================\n",
    "FAIL: test_eat (__main__.Testactivity)\n",
    "----------------------------------------------------------------------\n",
    "Traceback (most recent call last):\n",
    "  File \"unittest_testing_class.py\", line 10, in test_eat\n",
    "    \"veg food is healthy\"\n",
    "AssertionError: None != 'veg food is healthy'\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 1 test in 0.000s\n",
    "\n",
    "FAILED (failures=1)\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py\n",
    "F\n",
    "======================================================================\n",
    "FAIL: test_eat (__main__.Testactivity)\n",
    "----------------------------------------------------------------------\n",
    "Traceback (most recent call last):\n",
    "  File \"unittest_testing_class.py\", line 10, in test_eat\n",
    "    \"veg food is healthy\"\n",
    "AssertionError: None != 'veg food is healthy'\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 1 test in 0.000s\n",
    "\n",
    "FAILED (failures=1)\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py\n",
    ".\n",
    "----------------------------------------------------------------------\n",
    "Ran 1 test in 0.000s\n",
    "\n",
    "OK\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py\n",
    ".\n",
    "----------------------------------------------------------------------\n",
    "Ran 1 test in 0.000s\n",
    "\n",
    "OK\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py\n",
    ".\n",
    "----------------------------------------------------------------------\n",
    "Ran 1 test in 0.000s\n",
    "\n",
    "OK\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py\n",
    "F\n",
    "======================================================================\n",
    "FAIL: test_eat (__main__.Testactivity)\n",
    "----------------------------------------------------------------------\n",
    "Traceback (most recent call last):\n",
    "  File \"unittest_testing_class.py\", line 10, in test_eat\n",
    "    \"veg food is healthy\"\n",
    "AssertionError: 'lol' != 'veg food is healthy'\n",
    "- lol\n",
    "+ veg food is healthy\n",
    "\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 1 test in 0.000s\n",
    "\n",
    "FAILED (failures=1)\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py\n",
    "F\n",
    "======================================================================\n",
    "FAIL: test_eat (__main__.Testactivity)\n",
    "----------------------------------------------------------------------\n",
    "Traceback (most recent call last):\n",
    "  File \"unittest_testing_class.py\", line 10, in test_eat\n",
    "    \"veg food is healthy\"\n",
    "AssertionError: 'lol' != 'veg food is healthy'\n",
    "- lol\n",
    "+ veg food is healthy\n",
    "\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 1 test in 0.000s\n",
    "\n",
    "FAILED (failures=1)\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py\n",
    "F\n",
    "======================================================================\n",
    "FAIL: test_eat (__main__.Testactivity)\n",
    "----------------------------------------------------------------------\n",
    "Traceback (most recent call last):\n",
    "  File \"unittest_testing_class.py\", line 10, in test_eat\n",
    "    \"veg food is healthy\"\n",
    "AssertionError: 'lol' != 'veg food is healthy'\n",
    "- lol\n",
    "+ veg food is healthy\n",
    "\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 1 test in 0.000s\n",
    "\n",
    "FAILED (failures=1)\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py\n",
    "F\n",
    "======================================================================\n",
    "FAIL: test_eat (__main__.Testactivity)\n",
    "----------------------------------------------------------------------\n",
    "Traceback (most recent call last):\n",
    "  File \"unittest_testing_class.py\", line 10, in test_eat\n",
    "    \"veg food is healthy\"\n",
    "AssertionError: 'lol' != 'veg food is healthy'\n",
    "- lol\n",
    "+ veg food is healthy\n",
    "\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 1 test in 0.000s\n",
    "\n",
    "FAILED (failures=1)\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py\n",
    "F\n",
    "======================================================================\n",
    "FAIL: test_eat (__main__.Testactivity)\n",
    "----------------------------------------------------------------------\n",
    "Traceback (most recent call last):\n",
    "  File \"unittest_testing_class.py\", line 10, in test_eat\n",
    "    \"veg food is healthy\"\n",
    "AssertionError: 'lol' != 'veg food is healthy'\n",
    "- lol\n",
    "+ veg food is healthy\n",
    "\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 1 test in 0.000s\n",
    "\n",
    "FAILED (failures=1)\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 0 tests in 0.000s\n",
    "\n",
    "OK\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python demo_test.py\n",
    "F\n",
    "======================================================================\n",
    "FAIL: test_eat (__main__.DemoTest)\n",
    "----------------------------------------------------------------------\n",
    "Traceback (most recent call last):\n",
    "  File \"demo_test.py\", line 9, in test_eat\n",
    "    \"challi\"\n",
    "AssertionError: None != 'challi'\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 1 test in 0.001s\n",
    "\n",
    "FAILED (failures=1)\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python demo_test.py\n",
    "FF\n",
    "======================================================================\n",
    "FAIL: test_eat (__main__.DemoTest)\n",
    "----------------------------------------------------------------------\n",
    "Traceback (most recent call last):\n",
    "  File \"demo_test.py\", line 9, in test_eat\n",
    "    \"challi\"\n",
    "AssertionError: None != 'challi'\n",
    "\n",
    "======================================================================\n",
    "FAIL: test_uneat (__main__.DemoTest)\n",
    "----------------------------------------------------------------------\n",
    "Traceback (most recent call last):\n",
    "  File \"demo_test.py\", line 14, in test_uneat\n",
    "    \"tea\"\n",
    "AssertionError: None != 'tea'\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 2 tests in 0.001s\n",
    "\n",
    "FAILED (failures=2)\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py\n",
    "F\n",
    "======================================================================\n",
    "FAIL: test_eat (__main__.Testactivity)\n",
    "----------------------------------------------------------------------\n",
    "Traceback (most recent call last):\n",
    "  File \"unittest_testing_class.py\", line 10, in test_eat\n",
    "    \"veg food is healthy\"\n",
    "AssertionError: 'lol' != 'veg food is healthy'\n",
    "- lol\n",
    "+ veg food is healthy\n",
    "\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 1 test in 0.000s\n",
    "\n",
    "FAILED (failures=1)\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py\n",
    "F\n",
    "======================================================================\n",
    "FAIL: test_eat (__main__.Testactivity)\n",
    "----------------------------------------------------------------------\n",
    "Traceback (most recent call last):\n",
    "  File \"unittest_testing_class.py\", line 16, in test_eat\n",
    "    \"non veg is not healthy.\"\n",
    "AssertionError: 'non veg is not healthy' != 'non veg is not healthy.'\n",
    "- non veg is not healthy\n",
    "+ non veg is not healthy.\n",
    "?                       +\n",
    "\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 1 test in 0.001s\n",
    "\n",
    "FAILED (failures=1)\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py\n",
    "F\n",
    "======================================================================\n",
    "FAIL: test_eat (__main__.Testactivity)\n",
    "----------------------------------------------------------------------\n",
    "Traceback (most recent call last):\n",
    "  File \"unittest_testing_class.py\", line 16, in test_eat\n",
    "    \"lol.\"\n",
    "AssertionError: 'lol' != 'lol.'\n",
    "- lol\n",
    "+ lol.\n",
    "?    +\n",
    "\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 1 test in 0.001s\n",
    "\n",
    "FAILED (failures=1)\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py\n",
    ".\n",
    "----------------------------------------------------------------------\n",
    "Ran 1 test in 0.000s\n",
    "\n",
    "OK\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py -v\n",
    "test_eat (__main__.Testactivity) ... ok\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 1 test in 0.000s\n",
    "\n",
    "OK\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py -v\n",
    "test_check_num (__main__.Testactivity) ... ok\n",
    "test_eat (__main__.Testactivity) ... ok\n",
    "test_num (__main__.Testactivity) ... ok\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 3 tests in 0.000s\n",
    "\n",
    "OK\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py -v\n",
    "test_check_num (__main__.Testactivity)\n",
    "this is 3 ... ok\n",
    "test_eat (__main__.Testactivity)\n",
    "this is non veg ... ok\n",
    "test_num (__main__.Testactivity)\n",
    "this is 1 ... ok\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 3 tests in 0.000s\n",
    "\n",
    "OK\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py -v\n",
    "test_check_num (__main__.Testactivity)\n",
    "this is 3 ... ok\n",
    "test_eat (__main__.Testactivity)\n",
    "this is non veg ... ok\n",
    "test_num (__main__.Testactivity)\n",
    "this is 1 ... ok\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 3 tests in 0.000s\n",
    "\n",
    "OK\n",
    "rahish@rahish-Lenovo-ideapad-320-15ISK:~/Desktop/30_days_code_challange$ python3 unittest_testing_class.py -v\n",
    "test_check_num (__main__.Testactivity)\n",
    "this is 3 ... ok\n",
    "test_eat (__main__.Testactivity)\n",
    "this is non veg ... ok\n",
    "test_eats (__main__.Testactivity)\n",
    "this is veg ... ok\n",
    "test_num (__main__.Testactivity)\n",
    "this is 1 ... ok\n",
    "\n",
    "----------------------------------------------------------------------\n",
    "Ran 4 tests in 0.001s\n",
    "\n",
    "OK"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Before and After Hooks\n",
    "\n",
    "## setUp and tearDown\n",
    "\n",
    "For larger applications, you may want similar application state before running tests<br>\n",
    "setUp runs before each test method<br>\n",
    "tearDown runs after each test method<br>\n",
    "Common use cases: adding/removing data from a test database, creating instances of a class<br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Example\n",
    "\n",
    "class SomeTests(unittest.TestCase):\n",
    "\n",
    "    def setUp(self):\n",
    "        # do setup here\n",
    "        pass\n",
    "\n",
    "    def test_first(self):\n",
    "        # setUp runs before\n",
    "        # tearDown runs after\n",
    "        pass\n",
    "\n",
    "    def test_second(self):\n",
    "        # setUp runs before\n",
    "        # tearDown runs after\n",
    "        pass\n",
    "\n",
    "    def tearDown(self):\n",
    "        # do teardown here\n",
    "        pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Recap\n",
    "\n",
    "Tests help streamline development and reduce bugs<br>\n",
    "You can start with tests if doing TDD / Red, Green, Refactor<br>\n",
    "You can perform simple checks with assert<br>\n",
    "You can test with doctests, but typically shouldn't<br>\n",
    "unittest is a feature-rich, OOP style testing library in Python<br>\n",
    "To reduce code duplication in tests, use before/after hooks!<br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
